File: apps/backend/ent/schema/experience.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
	"github.com/google/uuid"
)

type Experience struct {
	ent.Schema
}

func (Experience) Mixin() []ent.Mixin {
	return []ent.Mixin{
		BaseMixin{},
	}
}

func (Experience) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("user_id", uuid.UUID{}).
			Comment("References auth.users(id)"),
		field.String("title").
			NotEmpty().
			MaxLen(200).
			Comment("Experience title"),
		field.String("category").
			Optional().
			MaxLen(50).
			Comment("Activity type: 인턴, 대외활동, 프로젝트, 아르바이트"),
		field.Time("period_start").
			Optional().
			Nillable().
			Comment("Start date"),
		field.Time("period_end").
			Optional().
			Nillable().
			Comment("End date"),
		field.String("role").
			Optional().
			MaxLen(100).
			Comment("Role in the experience"),
		field.Text("content").
			NotEmpty().
			Comment("Detailed experience content"),
		field.Text("result").
			Optional().
			Comment("Outcome / achievements"),
		field.Text("star_situation").
			Optional().
			Comment("STAR: Situation"),
		field.Text("star_task").
			Optional().
			Comment("STAR: Task"),
		field.Text("star_action").
			Optional().
			Comment("STAR: Action"),
		field.Text("star_result").
			Optional().
			Comment("STAR: Result"),
		field.JSON("keywords", []string{}).
			Optional().
			Comment("Core keywords array"),
		// embedding VECTOR(1536) — handled via raw SQL migration
		field.String("source").
			Default("manual").
			MaxLen(20).
			Comment("Input method: manual, interview"),
		field.Bool("is_archived").
			Default(false).
			Comment("Whether experience is archived"),
	}
}

func (Experience) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("user", UserProfile.Type).
			Ref("experiences").
			Unique().
			Required().
			Field("user_id").
			Annotations(entsql.OnDelete(entsql.Cascade)),
		edge.To("tags", ExperienceTag.Type),
		edge.To("weapons", ExperienceWeapon.Type),
		edge.To("usages", ExperienceUsage.Type),
	}
}

func (Experience) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("user_id"),
		index.Fields("user_id", "category"),
		index.Fields("user_id", "created_at"),
		index.Fields("user_id", "is_archived"),
	}
}

File: apps/web/src/lib/validations/experience.ts
import { z } from "zod/v4";

export const experienceCategories = [
  "인턴",
  "대외활동",
  "프로젝트",
  "아르바이트",
  "동아리",
  "봉사활동",
  "기타",
] as const;

export type ExperienceCategory = (typeof experienceCategories)[number];

export const experienceSchema = z
  .object({
    title: z
      .string()
      .min(2, "제목은 최소 2자 이상이어야 합니다.")
      .max(100, "제목은 100자 이내로 입력해주세요."),
    category: z.string().optional(),
    period_start: z.string().optional(),
    period_end: z.string().optional(),
    role: z.string().max(50, "역할은 50자 이내로 입력해주세요.").optional(),
    star_situation: z
      .string()
      .max(1000, "Situation은 1000자 이내로 입력해주세요.")
      .optional(),
    star_task: z
      .string()
      .max(1000, "Task는 1000자 이내로 입력해주세요.")
      .optional(),
    star_action: z
      .string()
      .max(2000, "Action은 2000자 이내로 입력해주세요.")
      .optional(),
    star_result: z
      .string()
      .max(1000, "Result는 1000자 이내로 입력해주세요.")
      .optional(),
    content: z
      .string()
      .max(5000, "자유 입력은 5000자 이내로 입력해주세요.")
      .optional(),
  })
  .refine(
    (data) => {
      if (data.period_start && data.period_end) {
        return new Date(data.period_end) >= new Date(data.period_start);
      }
      return true;
    },
    {
      message: "종료일은 시작일 이후여야 합니다.",
      path: ["period_end"],
    }
  );

export type ExperienceFormValues = z.infer<typeof experienceSchema>;
